File: LeetCode/5. 最长回文子串.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
  if (s.length < 2) return s;
  let origin = s;
  s = addSplit(s, "#");

  let begin = 0,
    maxLen = 1;
  let maxRight = 0,
    center = 0;
  let dp = Array(s.length).fill(0);
  for (let i = 0; i < s.length; i++) {
    if (i < maxRight) {
      let mirror = 2 * center - i;
      dp[i] = Math.min(dp[mirror], maxRight - i);
    }
    let l = i - dp[i] - 1,
      r = i + dp[i] + 1;
    while (l >= 0 && r < s.length && s[l] === s[r]) dp[i]++, l--, r++;
    if (i + dp[i] > maxRight) {
      center = i;
      maxRight = i + dp[i];
    }
    if (dp[i] > maxLen) {
      maxLen = dp[i];
      begin = (i - maxLen) / 2;
    }
  }
  return origin.substr(begin, maxLen);
};

const addSplit = (s, sp) => s.replace(/./g, "#$&") + "#";

File: LeetCode/5. 最长回文子串.cpp
class Solution
{
private:
  string addSplit(string &s, string sp)
  {
    string ss = "#";
    for (char c : s)
    {
      ss += c;
      ss += "#";
    }
    return ss;
  }

public:
  string longestPalindrome(string s)
  {
    if (s.length() < 2)
      return s;
    string ss = addSplit(s, "#");
    int maxLen = 1, begin = 0;
    int center = 0, maxRight = 0;
    vector<int> dp(ss.length(), 0);
    for (int i = 0; i < ss.length(); i++)
    {
      if (i < maxRight)
      {
        int mirror = 2 * center - i;
        dp[i] = min(dp[mirror], maxRight - i);
      }
      int l = i - dp[i] - 1, r = i + dp[i] + 1;
      while (l >= 0 && r < ss.length() && ss[l] == ss[r])
        l--, r++, dp[i]++;
      if (i + dp[i] > maxRight)
      {
        maxRight = i + dp[i];
        center = i;
      }
      if (dp[i] > maxLen)
      {
        maxLen = dp[i];
        begin = (i - maxLen) / 2;
      }
    }

    return s.substr(begin, maxLen);
  }
};